### PR TITLE
cocoa: option to scale window by HiDPI scale factor

### PR DESCRIPTION
**Pull request by Akemi**
thought it would probably better to discuss this here instead of the issue.

### CODE REVIEW

**Comment by atomnuker** [comment]
For simplicity, couldn't you just do the same thing we do in Wayland - always scale the window? If someone wanted to disable it they'd just have to adjust window-scale, since that's what the option is there for.

We wouldn't be able to reuse the hidpi scaling option for Wayland anyway since if we don't signal to the compositor we support scaling the compositor (currently only mutter but others might do the same in the future) will apply bilinear scaling and make everything blurry. So to really disable it we'd have to mess with the window-scale property and run the same code anyway, which would be a hack.

**Comment by Akemi** [comment]
i don't think always scaling the window is an option. current behaviour is actually what most people expect and Apple wants. personally i don't really want to change expected behaviour so the majority has to disable it.

lets see if i understand your suggestion right. assuming we open a 1280x720 video on macOS.
```
window size (px)		normal res		HiDPI res
current/PR (enabled):	1280x720		2560x1440
with PR (disabled):		1280x720		1280x720
with scaling:			1280x720		1280x720
with scaling (2x):		2560x1440		2560x1440
```
so if i want to change it back to our current behaviour with the scaling approach, it would be impossible to do so with just setting the window scale (2x). it would need some logic to differentiate between HiDPI and normal resolutions, which would make the code more bloated i believe.

i also didn't really plan for this option to be reused, since Apple's DPI scaling seems to work fundamentally different to windows' and linux' DPI scaling it seems. which makes it kinda impossible to make it work similar on every platform without violating some guidelines or expected behaviour.

**Comment by wm4** [comment]
(I don't mind.)

**Comment by haasn** [comment]
My issue with OS X and HighDPI is that apparently sometimes, it will take the output of the mpv window and further scale it using its own algorithm?

This should be avoided at all costs, ideally every pixel output by mpv should become a pixel on-screen.

**Comment by Akemi** [comment]
i am going to try explaining the behaviour in the next paragraphs. no need to read everything here a TL;DR, we already do everything we can so OSX isn't scaling anything. everything else is beyond our possibilities.

depending on your settings it will downscale (or upscale?) the desktop. best i can describe it, OSX uses a 'virtual' desktop resolution. normal HiDPI modes are usually tagged for example [1920x1080 HiDPI](http://forums.macrumors.com/attachments/display-png.407164/) which just means render in double the resolution (3840x2160) and render all elements in twice the size as well. so a 2x scale, just rendered in the actual res. so 1920x1080 HiDPI just means double the res but it should appear like 1080p in 'size'. usually you just use [Best for Display](http://forums.macrumors.com/attachments/os-x-101002-systempreferencesdisplays-scaling-1920x1200-a-jpg.515555/) option and it will automatically use the native resolution with 2x element assets for HiDPI. so no scaling whatsoever is done. only in the case if the 2x image assets are missing, but then only those images are scaled up. though that's not a problem concerning mpv.

then there is that case of using "Scaled" as seen in the previous image. "More Space" means render the Desktop in a higher resolution than the native res of the display. i don't have that scaled option on my hackintosh so i am just picking some theoretical numbers. lets assume the same setup as above and assuming the outmost right slider position for "More Space" means 1.5x the resolution of the native one. the native res is 3840x2160 so our virtual desktop is rendered at 5760x4320, though at the same time it only renders all elements in twice the size (OSX doesn't support fractional scaling of elements, only natural number scaling). so now since our native res is 3840x2160 the 5760x4320 are being downscaled by the system to that native resolution. so "More Space" just means render at a higher resolution than "Best" so all elements appear smaller on the actual resolution.

i never really looked at the other case "Larger Text" but i just conclude it does the exact opposite. virtual desktop in a smaller resolution than native and render all elements 2x the size. after that upscaling the virtual resolution the the native one. that way all elements look bigger than in the "Best" case.

mpv itself only sees that virtual resolution and can only render to that virtual desktop. it already renders pixel exact, as in at the moment it renders in twice the size on HiDPI resolutions, if the window is created on a screen with a HiDPI res. though not in the case when it is moved from normal to HiDPI. so in the case of a HiDPI res, mpv scales the video content. this patch just changes the behaviour that it will display every content in the native resolution of the video, when mpv is started, not depending in the HiDPI settings.

so the scaling that OSX does can't be prevented. it only happens when not using the "Best" settings in the resolution preferences. it's similar to changing the resolution to a not native display resolution and the display itself handles scaling, just that in our case the system handles the actual scaling. we don't know what magical algorithms Apple uses and we can't change it. OSX is limited in its element scaling, as in it can only render elements at [X times the size of a natural number](https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html) (currently only 1.0 and 2.0). instead of scaling elements by a fractional number it changes the resolution of the 'virtual desktop' to scale elements relatively.

so it just boils down to: Apple's DPI scaling being limited, "use our System the fuck we want you to" and "we know best what is best for you".

the only possible solution, to prevent the systems scaling, would be to change the desktop resolution to the native one or the "Best" scaling. though mode changing should only be implemented as an option if at all or a user script if possible.

i hope this explanation helps some people to understand the underlying problem and limitations, and that it won't start a discussion about how 'dumb' this is. we can't change it anyway.

**Comment by steve3d** [comment]
well, I vote for this, this is the solution for my issues. and for a media player, playing 4K video scaled down under a real 4k monitor is really not acceptable.

**Comment by pigoz** [comment]
LGTM, if there are no complaints by tomorrow I'll merge.

File: options/options.c [hunk 1]
```diff
@@ -216,6 +217,7 @@ const struct m_sub_options vo_sub_opts = {
         .panscan = 0.0f,
         .keepaspect = 1,
         .keepaspect_window = 1,
+        .hidpi_window_scale = 1,
```
Comment (steve3d) [inline options/options.c]: why set default to yes? is it ok to set it default to no? I really want to not use window scale as default.

Comment (Akemi) [inline options/options.c]: see my second reply on the PR.

Comment (atomnuker) [inline options/options.c]: Yep, having it on by default makes more sense. Anyone who doesn't want it can just disable it.

**Comment by steve3d** [comment]
with this merged, my issue can be closed #3716

**Comment by Argon-** [comment]
You can close it yourself.

**Comment by pigoz** [comment]
Cherry picked into master (aceeeaf9bb)

**Pull request closed without merging**